src/wasm/wasm-emscripten.cpp

```diff
@@ -444,6 +479,111 @@ void EmscriptenGlueGenerator::replaceStackPointerGlobal() {
   wasm.removeGlobal(stackPointer->name);
 }
 
+struct StackLimitEnforcer : public PostWalker<StackLimitEnforcer> {
+  StackLimitEnforcer(Global* stackPointer,
+                     Global* stackLimit,
+                     Builder& builder,
+                     Name handler)
+    : stackPointer(stackPointer), stackLimit(stackLimit), builder(builder),
+      handler(handler) {}
+
+  void noteNonLinear(Expression* curr) {
+    // End of this basic block; clear sets.
+    sets.clear();
+  }
+
+  void visitLocalSet(LocalSet* curr) { sets[curr->index] = curr; }
+
+  // Checks if an expression is addition.
+  // If we are adding to the stack pointer, for example, in the function
+  // epilogue, overflow is impossible.
+  bool isAddition(Expression* arg) {
+    while (auto* get = arg->dynCast<LocalGet>()) {
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      }
```

Comment:
Infinite loop if `!set`!

Reply:
I agree with Alon that it's probably not necessary to have all this extra logic.